File: src/services/runtime-preflight.js
import { access } from "node:fs/promises";
import { constants as fsConstants } from "node:fs";
import path from "node:path";
import { spawnSync } from "node:child_process";
import { fileURLToPath } from "node:url";

const DEFAULT_SYSTEM_CONFIG_DIR = "/etc/proxyrelay";
const DEFAULT_SYSTEM_RUNTIME_DIR = "/var/lib/proxyrelay/runtime";
const DEFAULT_SYSTEMD_DIR = "/etc/systemd/system";

function normalizePath(input) {
  return input instanceof URL ? fileURLToPath(input) : String(input);
}

async function hasAccess(targetPath, mode = fsConstants.F_OK) {
  try {
    await access(targetPath, mode);
    return true;
  } catch {
    return false;
  }
}

function findBinaryInPath(binary) {
  const result = spawnSync("which", [binary], { encoding: "utf8" });
  if (result.status === 0) {
    return result.stdout.trim();
  }

  return "";
}

function createCheck(key, title, level, summary, recommendation = "") {
  return {
    key,
    title,
    ok: level !== "error",
    level,
    summary,
    recommendation
  };
}

function dedupeRecommendations(checks) {
  return [...new Set(checks.map((check) => check.recommendation).filter(Boolean))];
}

export async function buildRuntimePreflightSnapshot({
  configPath,
  config,
  layout,
  state,
  controller,
  serviceDir = DEFAULT_SYSTEMD_DIR,
  expectedSystemRuntimeDir = DEFAULT_SYSTEM_RUNTIME_DIR
}) {
  const normalizedConfigPath = normalizePath(configPath);
  const configDir = path.resolve(path.dirname(normalizedConfigPath));
  const systemConfig = configDir === DEFAULT_SYSTEM_CONFIG_DIR;
  const runtimeDir = path.resolve(layout.runtimeDir);
  const providersDir = path.resolve(layout.providersDir);
  const runtimeAligned = !systemConfig || runtimeDir === expectedSystemRuntimeDir;
  const runtimeWritable = await hasAccess(runtimeDir, fsConstants.W_OK);
  const providersWritable = await hasAccess(providersDir, fsConstants.W_OK);
  const renderedConfigExists = await hasAccess(layout.mihomoConfigPath);
  const lastGoodConfigExists = await hasAccess(layout.lastGoodConfigPath);
  const stateFileExists = await hasAccess(layout.statePath);
  const systemctlAvailable = Boolean(findBinaryInPath("systemctl"));
  const proxyrelayUnitPath = path.join(serviceDir, "proxyrelayd.service");
  const mihomoUnitPath = path.join(serviceDir, "mihomo.service");
  const proxyrelayUnitExists = await hasAccess(proxyrelayUnitPath);
  const mihomoUnitExists = await hasAccess(mihomoUnitPath);
  const renderOnly = Boolean(config.runtime.render_only);
  const binaryFound = Boolean(state.runtime?.mihomoBinaryFound);
  const controllerReachable = Boolean(controller?.reachable);
  const controllerSecretConfigured = Boolean(config.runtime.external_secret);
  const applyStatus = state.runtime?.lastApplyStatus || "idle";
  const applyMessage = state.runtime?.lastApplyMessage || "尚未执行运行态下发";

  const checks = [
    createCheck(
      "runtime-path",
      "运行目录对齐",
      runtimeAligned ? "success" : "error",
      runtimeAligned
        ? `当前运行目录为 ${runtimeDir}`
        : `当前运行目录 ${runtimeDir} 与 Ubuntu 部署目标 ${expectedSystemRuntimeDir} 不一致`,
      runtimeAligned ? "" : "将 runtime.workdir 调整为 /var/lib/proxyrelay/runtime"
    ),
    createCheck(
      "runtime-permissions",
      "运行目录权限",
      runtimeWritable && providersWritable ? "success" : "error",
      runtimeWritable && providersWritable
        ? "runtime 与 providers 目录可写"
        : "runtime 或 providers 目录不可写，控制面无法持续生成配置与状态文件",
      runtimeWritable && providersWritable ? "" : "检查 /var/lib/proxyrelay/runtime 目录权限并确保 proxyrelay 用户可写"
    ),
    createCheck(
      "rendered-config",
      "渲染产物",
      renderedConfigExists && lastGoodConfigExists && stateFileExists ? "success" : "error",
      renderedConfigExists && lastGoodConfigExists && stateFileExists
        ? "mihomo.yaml、last-good 与 control-state 已生成"
        : "渲染产物不完整，控制面还未形成稳定运行态",
      renderedConfigExists && lastGoodConfigExists && stateFileExists ? "" : "先启动控制面生成 runtime 目录下的三类核心文件"
    ),
    createCheck(
      "mihomo-binary",
      "Mihomo 二进制",
      binaryFound ? "success" : renderOnly ? "warn" : "error",
      binaryFound
        ? `已检测到 Mihomo 二进制：${state.runtime?.mihomoBinary || config.runtime.mihomo_binary}`
        : renderOnly
        ? "当前为 render-only 模式，允许暂时未安装 Mihomo"
        : "managed 模式下未检测到 Mihomo 二进制",
      binaryFound || renderOnly ? "" : `安装 Mihomo 并确认 ${config.runtime.mihomo_binary} 可执行`
    ),
    createCheck(
      "controller-secret",
      "Controller 密钥",
      controllerSecretConfigured ? "success" : renderOnly ? "info" : "warn",
      controllerSecretConfigured ? "已配置 external controller 密钥" : "当前未配置 external controller 密钥",
      controllerSecretConfigured || renderOnly ? "" : "为 Mihomo external-controller 配置 secret，并同步到 runtime.external_secret"
    ),
    createCheck(
      "controller-reachability",
      "Controller 连通性",
      renderOnly ? "info" : controllerReachable ? "success" : "error",
      renderOnly
        ? "当前为 render-only 模式，未要求 controller 可达"
        : controllerReachable
        ? "external controller 可达"
        : `external controller 不可达：${controller?.message || "未知错误"}`,
      renderOnly || controllerReachable ? "" : "确认 mihomo 已启动、external-controller 地址正确且 secret 一致"
    ),
    createCheck(
      "runtime-apply",
      "运行态下发",
      renderOnly ? "info" : applyStatus === "success" ? "success" : applyStatus === "failed" ? "error" : "warn",
      renderOnly ? "当前为 render-only 模式，仅生成配置文件" : applyMessage,
      renderOnly || applyStatus === "success" ? "" : "保存配置或执行重载后，检查 controller 是否真正加载了最新 mihomo.yaml"
    ),
    createCheck(
      "systemd-units",
      "systemd 服务",
      systemConfig
        ? systemctlAvailable && proxyrelayUnitExists && mihomoUnitExists
          ? "success"
          : "error"
        : "info",
      systemConfig
        ? systemctlAvailable && proxyrelayUnitExists && mihomoUnitExists
          ? "proxyrelayd 与 mihomo 的 service 文件已就位"
          : "systemd service 文件缺失，Ubuntu 常驻部署尚未完成"
        : "当前使用仓库内配置，跳过系统级 service 检查",
      systemConfig && (!systemctlAvailable || !proxyrelayUnitExists || !mihomoUnitExists)
        ? "执行 deploy/install-ubuntu.sh，或手动安装 proxyrelayd.service 与 mihomo.service"
        : ""
    )
  ];

  const ready = checks.every((check) => check.level !== "error");
  const recommendations = dedupeRecommendations(checks);
  const errors = checks.filter((check) => check.level === "error").length;
  const warnings = checks.filter((check) => check.level === "warn").length;
  const status = errors > 0 ? "fail" : warnings > 0 ? "warn" : "ok";

  return {
    generatedAt: new Date().toISOString(),
    mode: renderOnly ? "render-only" : "managed",
    status,
    ready,
    summary: {
      ready,
      total: checks.length,
      passed: checks.filter((check) => check.level === "success").length,
      warnings,
      errors,
      info: checks.filter((check) => check.level === "info").length
    },
    paths: {
      configPath: normalizedConfigPath,
      runtimeDir,
      providersDir,
      mihomoConfigPath: layout.mihomoConfigPath,
      lastGoodConfigPath: layout.lastGoodConfigPath,
      statePath: layout.statePath,
      proxyrelayServicePath: proxyrelayUnitPath,
      mihomoServicePath: mihomoUnitPath
    },
    checks,
    recommendations
  };
}

export function formatRuntimePreflightForCli(snapshot) {
  const lines = [
    "ProxyRelay Runtime Preflight",
    `模式: ${snapshot.mode}`,
    `就绪状态: ${snapshot.ready ? "通过" : "未通过"}`,
    `统计: 成功 ${snapshot.summary.passed} / 警告 ${snapshot.summary.warnings} / 错误 ${snapshot.summary.errors}`
  ];

  for (const check of snapshot.checks) {
    const icon =
      check.level === "success" ? "OK" : check.level === "warn" ? "WARN" : check.level === "error" ? "ERR" : "INFO";
    lines.push(`${icon} ${check.title}: ${check.summary}`);
  }

  if (snapshot.recommendations.length) {
    lines.push("建议操作:");
    for (const recommendation of snapshot.recommendations) {
      lines.push(`- ${recommendation}`);
    }
  }

  return `${lines.join("\n")}\n`;
}

File: internal/axis/runtime_preflight.go
package axis

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
)

const (
	defaultSystemConfigDir = "/etc/proxyrelay"
	defaultSystemdDir      = "/etc/systemd/system"
)

func createCheck(key, title, level, summary, recommendation string) RuntimeCheck {
	return RuntimeCheck{
		Key:            key,
		Title:          title,
		Level:          level,
		OK:             level != "error",
		Summary:        summary,
		Recommendation: recommendation,
	}
}

func BuildRuntimePreflightSnapshot(configPath string, config *Config, layout *RuntimeLayout, state *AppState, controller *ControllerState) RuntimePreflightSnapshot {
	systemConfig := filepath.Clean(filepath.Dir(configPath)) == defaultSystemConfigDir
	runtimeAligned := !systemConfig || filepath.Clean(layout.RuntimeDir) == defaultSystemRuntimeWorkdir
	systemctlAvailable := exec.Command("which", "systemctl").Run() == nil
	proxyrelayServicePath := filepath.Join(defaultSystemdDir, "proxyrelayd.service")
	mihomoServicePath := filepath.Join(defaultSystemdDir, "mihomo.service")
	controllerSecretConfigured := config.Runtime.ExternalSecret != ""
	renderOnly := config.Runtime.RenderOnly

	checks := []RuntimeCheck{
		createCheck("runtime-path", "运行目录对齐", ternaryLevel(runtimeAligned, "success", "error"),
			ternaryString(runtimeAligned, fmt.Sprintf("当前工作目录是 %s", layout.RuntimeDir), fmt.Sprintf("当前工作目录 %s 与推荐目录 %s 不一致", layout.RuntimeDir, defaultSystemRuntimeWorkdir)),
			ternaryString(runtimeAligned, "", "建议把工作目录调整为 /var/lib/proxyrelay/runtime。")),
		createCheck("runtime-permissions", "运行目录权限", ternaryLevel(isWritable(layout.RuntimeDir) && isWritable(layout.ProvidersDir), "success", "error"),
			ternaryString(isWritable(layout.RuntimeDir) && isWritable(layout.ProvidersDir), "工作目录和订阅目录都可以正常写入", "工作目录或订阅目录不可写，程序无法持续保存配置和状态"),
			ternaryString(isWritable(layout.RuntimeDir) && isWritable(layout.ProvidersDir), "", "请检查 /var/lib/proxyrelay/runtime 的目录权限，并确认运行用户有写入权限。")),
		createCheck("rendered-config", "渲染产物", ternaryLevel(fileExists(layout.MihomoConfigPath) && fileExists(layout.LastGoodConfigPath) && fileExists(layout.StatePath), "success", "error"),
			ternaryString(fileExists(layout.MihomoConfigPath) && fileExists(layout.LastGoodConfigPath) && fileExists(layout.StatePath), "代理核心配置、最近一次可用配置和状态文件都已生成", "程序需要的关键文件还没生成完整"),
			ternaryString(fileExists(layout.MihomoConfigPath) && fileExists(layout.LastGoodConfigPath) && fileExists(layout.StatePath), "", "请先启动服务，让程序生成工作目录中的关键文件。")),
	}

	binaryFound := state.Runtime.MihomoBinaryFound
	controllerReachable := controller != nil && controller.Reachable
	applyStatus := state.Runtime.LastApplyStatus
	applyMessage := firstNonEmpty(state.Runtime.LastApplyMessage, "尚未应用配置")

	mihomoLevel := "success"
	if !binaryFound {
		if renderOnly {
			mihomoLevel = "warn"
		} else {
			mihomoLevel = "error"
		}
	}
	checks = append(checks, createCheck("mihomo-binary", "代理核心程序", mihomoLevel,
		ternaryString(binaryFound, fmt.Sprintf("已检测到代理核心程序：%s", firstNonEmpty(state.Runtime.MihomoBinary, config.Runtime.MihomoBinary)), ternaryString(renderOnly, "当前只保存配置，暂时还没有安装代理核心程序", "已开启自动接管，但还没有找到代理核心程序")),
		ternaryString(binaryFound || renderOnly, "", fmt.Sprintf("请安装 Mihomo，并确认 %s 可以直接执行。", config.Runtime.MihomoBinary))))

	secretLevel := "success"
	if !controllerSecretConfigured {
		if renderOnly {
			secretLevel = "info"
		} else {
			secretLevel = "warn"
		}
	}
	checks = append(checks, createCheck("controller-secret", "代理核心访问密钥", secretLevel,
		ternaryString(controllerSecretConfigured, "已配置代理核心访问密钥", "当前还没有配置代理核心访问密钥"),
		ternaryString(controllerSecretConfigured || renderOnly, "", "请在 Mihomo 中设置访问密钥，并同步填写到 runtime.external_secret。")))

	controllerLevel := "success"
	if renderOnly {
		controllerLevel = "info"
	} else if !controllerReachable {
		controllerLevel = "error"
	}
	checks = append(checks, createCheck("controller-reachability", "代理核心连接", controllerLevel,
		ternaryString(renderOnly, "当前只保存配置，暂不要求连上代理核心", ternaryString(controllerReachable, "AXIS 已成功连上代理核心", fmt.Sprintf("AXIS 还没连上代理核心：%s", firstNonEmpty(controller.Message, "未知错误")))),
		ternaryString(renderOnly || controllerReachable, "", "请确认 Mihomo 已启动、连接地址填写正确，且访问密钥保持一致。")))

	applyLevel := "warn"
	if renderOnly {
		applyLevel = "info"
	} else if applyStatus == "success" {
		applyLevel = "success"
	} else if applyStatus == "failed" {
		applyLevel = "error"
	}
	checks = append(checks, createCheck("runtime-apply", "配置应用结果", applyLevel,
		ternaryString(renderOnly, "当前只生成配置文件，暂不会自动发送到代理核心", applyMessage),
		ternaryString(renderOnly || applyStatus == "success", "", "保存配置或点击应用后，请确认代理核心已经加载最新配置。")))

	systemdLevel := "info"
	if systemConfig {
		if systemctlAvailable && fileExists(proxyrelayServicePath) && fileExists(mihomoServicePath) {
			systemdLevel = "success"
		} else {
			systemdLevel = "error"
		}
	}
	checks = append(checks, createCheck("systemd-units", "开机自启动配置", systemdLevel,
		func() string {
			if !systemConfig {
				return "当前使用的是仓库内配置，暂不检查系统级常驻服务。"
			}
			if systemctlAvailable && fileExists(proxyrelayServicePath) && fileExists(mihomoServicePath) {
				return "AXIS 和 Mihomo 的系统服务文件都已就位。"
			}
			return "系统服务文件还不完整，当前还没有完成常驻部署。"
		}(),
		func() string {
			if systemConfig && (!systemctlAvailable || !fileExists(proxyrelayServicePath) || !fileExists(mihomoServicePath)) {
				return "请执行 deploy/install-ubuntu.sh，或手动安装 AXIS 与 Mihomo 的系统服务文件。"
			}
			return ""
		}()))

	recommendations := []string{}
	for _, check := range checks {
		if check.Recommendation != "" && !containsString(recommendations, check.Recommendation) {
			recommendations = append(recommendations, check.Recommendation)
		}
	}

	summary := RuntimePreflightSummary{Ready: true, Total: len(checks)}
	for _, check := range checks {
		switch check.Level {
		case "success":
			summary.Passed++
		case "warn":
			summary.Warnings++
		case "error":
			summary.Errors++
			summary.Ready = false
		case "info":
			summary.Info++
		}
	}
	status := "ok"
	if summary.Errors > 0 {
		status = "fail"
	} else if summary.Warnings > 0 {
		status = "warn"
	}

	return RuntimePreflightSnapshot{
		GeneratedAt: nowISO(),
		Mode:        ternaryString(renderOnly, "render-only", "managed"),
		Status:      status,
		Ready:       summary.Ready,
		Summary:     summary,
		Paths: RuntimePreflightPaths{
			ConfigPath:            configPath,
			RuntimeDir:            layout.RuntimeDir,
			ProvidersDir:          layout.ProvidersDir,
			MihomoConfigPath:      layout.MihomoConfigPath,
			LastGoodConfigPath:    layout.LastGoodConfigPath,
			StatePath:             layout.StatePath,
			ProxyrelayServicePath: proxyrelayServicePath,
			MihomoServicePath:     mihomoServicePath,
		},
		Checks:          checks,
		Recommendations: recommendations,
	}
}

func isWritable(dir string) bool {
	file, err := os.CreateTemp(dir, "axis-write-test-*")
	if err != nil {
		return false
	}
	file.Close()
	os.Remove(file.Name())
	return true
}

func ternaryString(condition bool, left, right string) string {
	if condition {
		return left
	}
	return right
}

func ternaryLevel(condition bool, left, right string) string {
	if condition {
		return left
	}
	return right
}

func containsString(items []string, target string) bool {
	for _, item := range items {
		if item == target {
			return true
		}
	}
	return false
}
